src/proxies/InverterTransparentUpgradeableProxy_v1.sol

```diff
@@ -0,0 +1,120 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity 0.8.23;
+
+//Internal Interfaces
+import {IInverterTransparentUpgradeableProxy_v1} from
+    "src/proxies/interfaces/IInverterTransparentUpgradeableProxy_v1.sol";
+import {IInverterBeacon_v1} from "src/proxies/interfaces/IInverterBeacon_v1.sol";
+
+// External Dependencies
+import {ERC1967Proxy} from "@oz/proxy/ERC1967/ERC1967Proxy.sol";
+import {ERC1967Utils} from "@oz/proxy/ERC1967/ERC1967Utils.sol";
+import {ERC165} from "@oz/utils/introspection/ERC165.sol";
+
+/**
+ * @title   Inverter TransparentUpgradeableProxy
+ *
+ * @notice  An alternative variant of the TransparentUpgradeableProxy of OpenZeppelin that allows for
+ *          upgrading the linked implementation of the proxy to the implementation provided by a linked beacon.
+ *
+ * @dev This contract is basically a copy of the TransparentUpgradeableProxy from openzeppelin.
+ *      We adapted the callable functions of the admin account to only be able to upgrade the implementation to the newest implementation of the linked InverterBeacon.
+ *
+ *      !!! IMPORTANT !!!
+ *
+ *      1. If any account other than the admin calls the proxy, the call will be forwarded to the implementation, even if
+ *      that call matches the {IInverterTransparentUpgradeableProxy_v1-upgradeToNewestVersion} function exposed by the proxy itself.
+ *      2. If the admin calls the proxy, it can call the `upgradeToNewestVersion` function but any other call won't be forwarded to
+ *      the implementation. If the admin tries to call a function on the implementation it will fail with an error indicating
+ *      the proxy admin cannot fallback to the target implementation.
+ *
+ *      These properties mean that the admin account can only be used for upgrading the proxy, so it's best if it's a
+ *      dedicated account that is not used for anything else. This will avoid headaches due to sudden errors when trying to
+ *      call a function from the proxy implementation. For this reason, the proxy deploys an instance of {ProxyAdmin} and
+ *      allows upgrades only if they come through it. You should think of the `ProxyAdmin` instance as the administrative
+ *      interface of the proxy, including the ability to change who can trigger upgrades by transferring ownership.
+ *
+ * @author  Inverter Network
+ */
+contract InverterTransparentUpgradeableProxy_v1 is ERC1967Proxy {
+    /// @notice The provided beacon address doesnt support the interface {IInverterBeacon_v1}
+    error InverterTransparentUpgradeableProxy__InvalidBeacon();
+
+    /// @dev If the proxy caller is the current admin then it can only call the admin functions.
+    error InverterTransparentUpgradeableProxy__ProxyDeniedAdminAccess();
+
+    //--------------------------------------------------------------------------------
+    // State
+
+    /// @dev The address of the admin that can update the implementation address of this proxy
+    address internal immutable _admin;
```

Comment:
In the original implementation of OZ this is `immutable` as it's using a ProxyAdmin. Here, it's probably an EOA, so this should be changeable here, right?

Reply:
I thought about this too.
The problem is that every single contract in the workflow would need to be updated for the admin to properly work.
And because only the admin can change the admin, then if the address should be corrupted the whole system would be useless anyway. The recommondation they gave in the Original proxy was to make it a specially created account in the first place so thats what I went with
Should I maybe add another disclaimer either in the workflowConfig or in the proxy itself?

Reply:
We agreed to make a seperate admin contract that allows to update the proxies via the immutable admin role
We will create a linear issue for this